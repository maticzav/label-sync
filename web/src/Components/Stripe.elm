module Components.Stripe exposing (stripe)

import Element exposing (..)
import Element.Background as Background
import Element.Border as Border
import Element.Font as Font
import Html.Attributes as Attr


labels : List ( String, Element.Color )
labels =
    [ ( "bug/0-needs-info", rgb255 238 0 0 )
    , ( "area/database", rgb255 251 202 4 )
    , ( "kind/improvement", rgb255 59 91 219 )
    , ( "bug/2-confirmed", rgb255 238 0 0 )
    , ( "kind/bug", rgb255 59 91 219 )
    , ( "area/docs", rgb255 251 202 4 )
    , ( "kind/feature", rgb255 59 91 219 )
    , ( "status/stale", rgb 36 218 141 )
    ]


stripe : Element msg
stripe =
    row
        [ width (fill |> maximum 700)
        , spacing 15
        , paddingXY 0 10
        , centerX
        , Element.htmlAttribute <| Attr.id "stripe"
        ]
        (List.map viewLabel labels)


viewLabel : ( String, Element.Color ) -> Element msg
viewLabel ( label, color ) =
    el
        [ Background.color color
        , Border.rounded 3
        , Font.color (rgb255 255 255 255)
        , Font.size 14
        , Font.bold
        , paddingXY 5 3
        , Font.family [ Font.sansSerif ]
        , Border.innerShadow
            { offset = ( 0.0, -0.4 )
            , size = 0.0
            , blur = 2.0
            , color = rgb255 27 31 35
            }
        ]
        (text label)
