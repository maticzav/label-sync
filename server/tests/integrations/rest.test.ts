import { PrismaClient, Plan } from '@prisma/client'
import nock from 'nock'
import { Probot, createProbot } from 'probot'
import request from 'request-promise-native'
import Stripe from 'stripe'
import { createLogger, Logger, format, transports } from 'winston'

const manager = require('../../src')

describe('rest:', () => {
  let prisma: PrismaClient
  let winston: Logger
  let stripe: Stripe

  let logs: string[] = []

  beforeAll(async () => {
    // Network settings
    nock.disableNetConnect()
    // local servers
    nock.enableNetConnect((host) => {
      return (
        host.includes('localhost') ||
        host.includes('127.0.0.1') ||
        host.includes('stripe')
      )
    })

    // DataStores
    prisma = new PrismaClient()
    winston = createLogger({
      level: 'info',
      exitOnError: false,
      format: format.json(),
      transports: [],
    })
    stripe = new Stripe(process.env.STRIPE_API_KEY!, {
      apiVersion: '2020-08-27',
    })

    /* Setup probot */
    probot = createProbot({
      id: 1,
      cert: 'test',
      githubToken: 'test',
      port: 4040,
    })

    probot.load((app) => manager(app, prisma, winston, stripe))
    probot.start()
  })

  afterAll(async () => {
    nock.cleanAll()
    nock.enableNetConnect()

    probot.httpServer!.close()
    await prisma.$disconnect()
  })

  let probot: Probot

  beforeEach(async () => {
    /* Clean the database */
    await prisma.installation.deleteMany({ where: {} })
  })

  afterEach(() => {
    nock.cleanAll()
  })

  /* Tests */

  test('FREE creates a resource', async () => {
    const body = await request({
      uri: `http://localhost:4040/subscribe/session`,
      method: 'POST',
      json: true,
      body: {
        email: 'test@label-sync.com',
        account: 'maticzav',
        agreed: true,
        plan: 'FREE',
      },
    }).promise()

    const installation = await prisma.installation.findOne({
      where: { account: 'maticzav' },
    })

    expect(body).toEqual({ status: 'ok', plan: 'FREE' })
    delete installation?.id
    delete installation?.createdAt
    delete installation?.updatedAt
    delete installation?.periodEndsAt
    expect(installation).toMatchSnapshot()
  })

  test('PAID creates a checkout session', async () => {
    const body = await request({
      uri: `http://localhost:4040/subscribe/session`,
      method: 'POST',
      json: true,
      body: {
        email: 'test@label-sync.com',
        account: 'maticzav',
        agreed: true,
        plan: 'PAID',
        period: 'ANNUALLY',
      },
    }).promise()

    expect(body.status).toBe('ok')

    /* Test Stripe */

    const session = await stripe.checkout.sessions.retrieve(body.session, {
      expand: ['subscription'],
    })

    expect(session.customer_email).toBe('test@label-sync.com')

    /* Test database */

    const installation = await prisma.installation.findOne({
      where: { account: 'maticzav' },
    })

    delete installation?.id
    delete installation?.createdAt
    delete installation?.updatedAt
    delete installation?.periodEndsAt
    expect(installation).toMatchSnapshot()
  })

  /* End of tests */
})
