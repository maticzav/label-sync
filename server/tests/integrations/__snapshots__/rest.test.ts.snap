// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rest: FREE creates a resource 1`] = `
Object {
  "account": "maticzav",
  "activated": false,
  "email": "test@label-sync.com",
  "plan": "FREE",
}
`;

exports[`rest: PAID creates a checkout session 1`] = `
Object {
  "account": "maticzav",
  "activated": false,
  "email": "test@label-sync.com",
  "plan": "FREE",
}
`;
